### PR TITLE
08 create an animal

### PR DESCRIPTION
**Pull request by afornalik**
Created "save" API endpoint to persist "Animal" entity.

### CODE REVIEW

**Comment by afornalik** [comment]
Hey Patrycja, please review following changes.

File: src/main/java/com/example/animalshelter/controller/AnimalController.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package com.example.animalshelter.controller;
+
+import com.example.animalshelter.model.Animal;
+import com.example.animalshelter.service.AnimalService;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController()
```
Comment (patrycja93) [inline src/main/java/com/example/animalshelter/controller/AnimalController.java:9]: Would do you think about switch it to:
`@RestController
@RequestMapping("/animals")
public class AnimalController {
`

[Here](https://spring.io/guides/tutorials/rest/) are tips about building rest.

File: src/main/java/com/example/animalshelter/controller/AnimalController.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+public class AnimalController {
+
+    private final AnimalService animalService;
+
+    public AnimalController(AnimalService animalService) {
+        this.animalService = animalService;
+    }
+
+    @PostMapping(value = "/save", consumes = "application/json", produces = "application/json")
```
Comment (patrycja93) [inline src/main/java/com/example/animalshelter/controller/AnimalController.java:18]: By using `RequestMapping(/animals)` on the class, we do not have to use `value=""` here.

File: src/main/java/com/example/animalshelter/controller/AnimalController.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    public Animal saveAnimal(@RequestBody Animal animal) {
```
Comment (patrycja93) [inline src/main/java/com/example/animalshelter/controller/AnimalController.java:19]: I would suggest changing the name from `save` to 'add`. What do you think? 
We do not save an animal we add it :)

File: src/main/java/com/example/animalshelter/controller/AnimalController.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        return animalService.save(animal) ? animal : new Animal();
```
Comment (patrycja93) [inline src/main/java/com/example/animalshelter/controller/AnimalController.java:20]: I think we should return the status code. Ok -201, fail - 400 ?

File: src/main/java/com/example/animalshelter/model/Animal.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * Keeps basic information about an animal
+ */
+@Getter
+@Setter
+@ToString
+public class Animal {
+
+    private int id;
```
Comment (patrycja93) [inline src/main/java/com/example/animalshelter/model/Animal.java:15]: I think this value should not be set by the user. Maybe should be static?

**Comment by patrycja93** [comment]
Please, look at my suggestions.

File: src/main/java/com/example/animalshelter/model/Animal.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+
+/**
+ * Keeps basic information about an animal
+ */
+@Getter
+@Setter
+@ToString
```
[the remark this answers is not in the document]
Comment (afornalik) [reply src/main/java/com/example/animalshelter/model/Animal.java:12]: ok, done.

File: src/main/java/com/example/animalshelter/model/Animal.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+public class Animal {
+
+    private int id;
+    private String name;
+    private String type;
```
[the remark this answers is not in the document]
Comment (afornalik) [reply src/main/java/com/example/animalshelter/model/Animal.java:17]: Done, added several basic types (dogs,cats,birds,..).

File: src/main/java/com/example/animalshelter/model/Animal.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    private String gender;
```
[the remark this answers is not in the document]
Comment (afornalik) [reply src/main/java/com/example/animalshelter/model/Animal.java:18]: Done

File: src/main/java/com/example/animalshelter/model/Animal.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    private int age;
+
+    // In the future we can switch this to enum to represent animal health status.
+    private String status;
```
[the remark this answers is not in the document]
Comment (afornalik) [reply src/main/java/com/example/animalshelter/model/Animal.java:22]: It something like is animal is health, sick, injured. Basically this describe health condition of an animal. I will change it to enum too.

File: src/main/java/com/example/animalshelter/controller/AnimalController.java (hunk 1 above)
Comment (afornalik) [reply src/main/java/com/example/animalshelter/controller/AnimalController.java:9]: done

File: src/main/java/com/example/animalshelter/controller/AnimalController.java (hunk 2 above)
Comment (afornalik) [reply src/main/java/com/example/animalshelter/controller/AnimalController.java:18]: ok, done.

File: src/main/java/com/example/animalshelter/controller/AnimalController.java (hunk 3 above)
Comment (afornalik) [reply src/main/java/com/example/animalshelter/controller/AnimalController.java:19]: yes, as we talk add will be better choice here, changed.

File: src/main/java/com/example/animalshelter/controller/AnimalController.java (hunk 4 above)
Comment (afornalik) [reply src/main/java/com/example/animalshelter/controller/AnimalController.java:20]: Now we return status 200 if request is resolved successfully and 400 if something go wrong, also there is string response.

File: src/main/java/com/example/animalshelter/model/Animal.java (hunk 5 above)
Comment (afornalik) [reply src/main/java/com/example/animalshelter/model/Animal.java:15]: Yes, agree we shouldn't set up id by us own. Probably it should be set up automatically by framework. We need to add specific annotation later to support that.

**Comment by afornalik** [comment]
Hi Patrycja, thank you for the code review. Please check the latest changes.

File: src/main/java/com/example/animalshelter/controller/AnimalController.java [hunk 10]
```diff
@@ -4,19 +4,24 @@
...
 import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RestController;
 
-@RestController()
+@RestController
+@RequestMapping("/animals")
 public class AnimalController {
 
+    private static final String SUCCESSFUL_RESPONSE = "Animal added successfully.";
```
Comment (patrycja93) [inline src/main/java/com/example/animalshelter/controller/AnimalController.java:14]: Proposition: The animal has been added successfully.

File: src/main/java/com/example/animalshelter/controller/AnimalController.java [hunk 11]
```diff
@@ -4,19 +4,24 @@
...
+    private static final String FAILED_RESPONSE = "An Error occurred during adding animal.";
+
```
Comment (patrycja93) [inline src/main/java/com/example/animalshelter/controller/AnimalController.java:16]: Please change the uppercase "E" to a lower one.

File: src/main/java/com/example/animalshelter/model/Animal.java (hunk 5 above)
Comment (patrycja93) [reply src/main/java/com/example/animalshelter/model/Animal.java:11]: Ok.

File: src/main/java/com/example/animalshelter/model/AnimalGender.java [hunk 12]
```diff
@@ -0,0 +1,8 @@
+package com.example.animalshelter.model;
+
+public enum AnimalGender {
+
+    UNKNOWN,
```
Comment (patrycja93) [inline src/main/java/com/example/animalshelter/model/AnimalGender.java:5]: I think the UNKNOW value is not necessary ;)

File: src/main/java/com/example/animalshelter/model/AnimalHealthStatus.java [hunk 13]
```diff
@@ -0,0 +1,11 @@
+package com.example.animalshelter.model;
+
+public enum AnimalHealthStatus {
+
+    UNKNOWN,
```
Comment (patrycja93) [inline src/main/java/com/example/animalshelter/model/AnimalHealthStatus.java:5]: Maybe a better option would be "not examined". Unknow says nothing.

File: src/main/java/com/example/animalshelter/service/AddAnimalException.java [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.BAD_REQUEST)
+public class AddAnimalException extends RuntimeException {
+
+    public AddAnimalException(String string) {
+        super("Could not create an animal. " + string);
+    }
```
Comment (patrycja93) [inline src/main/java/com/example/animalshelter/service/AddAnimalException.java:11]: Proposition: "Could not create the animal " + string + "."

File: src/main/java/com/example/animalshelter/service/AnimalServiceImpl.java [hunk 15]
```diff
@@ -16,10 +16,10 @@ public AnimalServiceImpl(AnimalRepository animalRepository) {
     }
 
     @Override
-    public boolean save(Animal animal) {
-        if (animal == null || animal.getId() <= 0) {
-            return false;
+    public boolean add(Animal animal) {
+        if (!animalRepository.add(animal)) {
+            throw new AddAnimalException("Error during saving an animal in database.");
```
Comment (patrycja93) [inline src/main/java/com/example/animalshelter/service/AnimalServiceImpl.java:21]: Please make this text as a `const`.

File: src/main/java/com/example/animalshelter/controller/AnimalController.java (hunk 10 above)
Comment (afornalik) [reply src/main/java/com/example/animalshelter/controller/AnimalController.java:14]: Done.

File: src/main/java/com/example/animalshelter/controller/AnimalController.java (hunk 11 above)
Comment (afornalik) [reply src/main/java/com/example/animalshelter/controller/AnimalController.java:16]: Done.

File: src/main/java/com/example/animalshelter/model/AnimalGender.java (hunk 12 above)
Comment (afornalik) [reply src/main/java/com/example/animalshelter/model/AnimalGender.java:5]: yes, this was an option when an user don't want to provide all details initially(like a default value). I will remove it.

File: src/main/java/com/example/animalshelter/model/AnimalHealthStatus.java (hunk 13 above)
Comment (afornalik) [reply src/main/java/com/example/animalshelter/model/AnimalHealthStatus.java:5]: That's right, changed.

File: src/main/java/com/example/animalshelter/service/AddAnimalException.java (hunk 14 above)
Comment (afornalik) [reply src/main/java/com/example/animalshelter/service/AddAnimalException.java:11]: Done.

File: src/main/java/com/example/animalshelter/service/AnimalServiceImpl.java (hunk 15 above)
Comment (afornalik) [reply src/main/java/com/example/animalshelter/service/AnimalServiceImpl.java:21]: Done.

**Comment by afornalik** [comment]
Thank you for the review. Now it looks better.

**Review by patrycja93 — APPROVED** [review]
LGTM :)

**Pull request merged**